apacheconfig/wloader.py

```diff
@@ -0,0 +1,219 @@
+#
+import abc
+import six
+
+from apacheconfig import make_parser
+from apacheconfig import make_lexer
+
+
+def _restore_original(word):
+    """If the `word` is a Quoted string, restores it to original.
+    """
+    if getattr(word, 'is_single_quoted', False):
+        return "'%s'" % word
+    if getattr(word, 'is_double_quoted', False):
+        return '"%s"' % word
+    return word
+
+
+@six.add_metaclass(abc.ABCMeta)
+class Node():
+    """Generic class containing data that represents a node in the config AST.
+    """
+
+    @abc.abstractmethod
+    def __str__(self):
+        """Writes this node to a raw string. To get more metadata about the
+        object, use ``repr``.
+        """
+        pass
+
+    @abc.abstractmethod
+    def __repr__(self):
+        """Returns a string containing object metadata."""
+        pass
+
+    @property
+    def parser_type(self):
+        """A typestring as defined by the apacheconfig parser.
+
+        The possible values for this are:
+
+        :class:`apacheconfig.ItemNode`: ``comment``, ``statement``,
+        ``include``, ``includeoptional``
+
+        :returns: The typestring (as defined by the apacheconfig parser) for
+                  this node.
+        :rtype: `str`
+        """
+        if self._type is None:
+            raise NotImplementedError()
+        return self._type
+
+    @property
+    def whitespace(self):
+        """A string representing literal trailing or preceding whitespace
+        for this node. Can be overwritten.
+
+        Each ``ItemNode`` or ``BlockNode`` keeps track of the whitespace
+        preceding it. For the first element in the configuration file, there
+        could be no whitespace preceding it at all, in which case this should
+        return the empty string.
+
+        ContentsNode is special in that it keeps track of the *trailing*
+        whitespace. For example::
+
+            ItemNode('\\n  option value').whitespace => "\\n  "
+            ItemNode('\\n  # comment').whitespace => "\\n  "
+            BlockNode('\\n  <a>\\n</a>').whitespace => "\\n  "
+            ContentsNode('\\n  option value # comment\\n').whitespace => "\\n"
+
+        :returns: a string containing literal preceding or trailing whitespace
+                  information for this node.
+        :rtype: `str`
+       """
+        if self._whitespace is None:
+            raise NotImplementedError()
+        return self._whitespace
+
+    @whitespace.setter
+    def whitespace(self, value):
+        """A string representing literal trailing or preceding whitespace
+        for this node. Trailing for ``Contents``, preceding for ``Item`` or
+        ``Block``.
+
+        :param str value: whitespace string to set this node's whitespace.
+        """
+        if self._whitespace is None:
+            raise NotImplementedError()
+        self._whitespace = value
+
+
+class ItemNode(Node):
+    """Contains data for a comment or option-value directive.
```

Comment:
I have a vague recollection that the first line of a docstring should be [benevolently imperative](https://www.python.org/dev/peps/pep-0257/).